### PR TITLE
Reading dockercfg files for authenticating with private registries

### LINKED ISSUE #660: Docker auth from docker config
I know secrets management is forthcoming, but it would also be nice if nomad could get docker auth from the default docker config files. 

The client could just attempt to call [NewAuthConfigurationsFromDockerCfg](https://godoc.org/github.com/fsouza/go-dockerclient#NewAuthConfigurationsFromDockerCfg). 

I'm trying to get nomad to pull images from an [AWS ECR](https://aws.amazon.com/ecr/), which requires obtaining a new auth password from the API every 12 hours. The CLI command `aws ecr get-login` returns a `docker login` command to be eval'ed in the shell, which in turn is written to `~/.docker/config.json`.

### PR DESCRIPTION
**Pull request by diptanu**
Fixed #660 and #702

### CODE REVIEW

**Comment by dadgar** [comment]
LGTM

**Comment by diptanu** [comment]
@dadgar Oh yeah #702 is related too. Updating the CHANGELOG too.

**Comment by trenton42** [comment]
This issue is bigger than just having the https:// on the beginning of the repo address (you should never have to hand edit the auths section of config.json -- that part is generated by `docker login`). Nomad also looks for the completely wrong thing when looking up auths.

In [docker.go](https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L476) it uses `repo` to look up the auth:

```go
if authConfiguration, ok := authConfigurations.Configs[repo]; ok {
    authOptions = authConfiguration
}
```

But `repo` is returned by [the docker client](https://github.com/fsouza/go-dockerclient/blob/master/misc.go#L48) and all it does is split the tag from the rest of the docker image path. What this means is that nomad is not just looking for the server address in the auths, it is also looking for _the image name_. For example, on GCE I might have an image called `gcr.io/some-project/some-image:v1.0`. If you want this to work right now, you have to put an auth key in for _ever single image you run_. When nomad tries to look up an auth in `config.json`, it will look for a key called `gcr.io/some-project/some-image`, but what `docker login` actually puts in `config.json` is:

```json
{
    "auths": {
        "https://gcro.io": {
            "auth": "<some token>",
            "email": "not@val.id"
        }
    }
}
```

What should happen is either:

1. Nomad uses `authOptions.ServerAddress` for the lookup (requires jobs to specify this in their config)
2. Parse the image name and determine the correct server name

TL;DR neither #660 nor #702 are fixed.

**Comment by waziers** [comment]
@diptanu why is this closed, @trenton42 is correct, this is still broken unless you manually re-write the `config.json` file

**Comment by trenton42** [comment]
@diptanu I think this did actually get fixed recently. The hitch I found is that it does look at the SSL flag in the docker driver config of the job, so I had to set that to true.

**Comment by diptanu** [comment]
@waziers @trenton42 So what we have implemented there is Nomad to be able to use the dockercfg file specified by an operator. Nomad doesn't automatically do `docker login` or anything special to re-generate the dockercfg file on ECS, GCP, Quay or any other vendor specific tool currently. We expect the operator to have automation in place  for re-generating the file.

The reason why this is hard to do is that the space is very fragmented and every registry provider has created their own way of authenticating with the registry and generate the credentials.

Also as @trenton42 has mentioned, if a registry is behind SSL, the `SSL` flag has to be set as true.
